llmfoundry/models/hf/hf_fsdp.py

```diff
@@ -135,17 +136,22 @@ def prepare_hf_causal_lm_model_for_fsdp(model: PreTrainedModel) -> None:
             raise ValueError(
                 f'Unable to FSDP-wrap this model! `{mod_name}` does not ' +
                 'follow common layer/weight naming conventions.')
+    
+    # For FSDP with models with different device intiailizations e.g. `cpu` and `meta`
+    # we need to tag all child modules that are torch.nn.Modules with `_fsdp_wrap`.
+    for child in model.children():
+        if isinstance(child, type(causal_base_model)):
+            continue
+        if isinstance(child, torch.nn.Module):
+            child._fsdp_wrap = True
+    
+    for child in causal_base_model.children():
+        if isinstance(child, torch.nn.ModuleList):
+            continue
+        if isinstance(child, torch.nn.Module):
+            child._fsdp_wrap = True
+    
     block_type = type(model_block[0])  # type: ignore
-    # When using the HF LM models,
-    # the weights of the self.lm_head and self.transformer.wte are tied.
-    # This tying occurs inside the `self.post_init()` function.
-    # This is a hurdle for FSDP because they need to be in the same FSDP block
-    # These lines ensures that both modules stay together in the top-most block when
-    # the model has this tying enabled (almost all do; this property defaults to True)
-    if model.config.tie_word_embeddings:
-        causal_base_model._fsdp_wrap = False  # type: ignore
-        tied_embeddings._fsdp_wrap = False  # type: ignore
-        lm_head._fsdp_wrap = False  # type: ignore
```

Comment:
I think getting rid of this entirely might be a bit heavy handed, especially if it breaks tied embeddings throughout the stack.

Instead, I recommend that we keep this logic for this PR and create a separate PR where we add in logic for mixed CPU/meta initialization. I feel like that separate PR would be the place where we would want to change around this `if` block to throw an informative error in the case that `tie_word_embeddings==True` and `tied_embeddings is not lm_head` and the user is attempting to do the mixed initialization.

Reply:
Yeah sounds good. I'll make a separate PR. I think we'd need the user to specify the mixed initialization flag, since otherwise we have no other way to know how the other models are initialized on other processes... Wdyt

Reply:
Yes, totally agree. I think the simplest thing would be to allow the HF model wrappers to use `cfg.init_device = "mixed"` (or whatever we want to call it). In a perfect world, this would be the default "cpu" initialization works when using FSDP and a compatible model, but detecting the latter condition might not be trivial.
Happy to discuss whether this is possible offline.